pilot/pkg/serviceregistry/kube/controller.go

```diff
@@ -740,7 +740,23 @@ func (c *Controller) AppendInstanceHandler(f func(*model.ServiceInstance, model.
 		if ep.Namespace == meta_v1.NamespaceSystem {
 			return nil
 		}
-		c.updateEDS(ep)
+
+		// Fix the scenario: when a proxy (Sidecar or gateway) connects to Pilot
+		// before Pilot receives the set of service instances associated with
+		// the service, from K8S. So env.GetProxyServiceInstances() call returns
+		// an empty set of service instances. This causes the Gateway code to
+		// fail, setting up the proxy with no listener.
+		if event == model.EventAdd {
```

Comment:
Did you test it? #10904 has steps to reproduce the issue of some gateway instances not getting the :80 listener.
When I played with it and reproduced it (following instructions there) I actually saw a combination of `EventUpdate` with empty ServiceInstances.

Reply:
I can not reproduce with this commit.  I think it fix the scenario described here https://github.com/istio/istio/pull/11905#issue-254777175

Reply:
The scenario in #10904 is due to the same problem. When you create more replicas of the gateways some of the proxies connects before the controller got event of the new instance. But I remember the events were an update and not add.
Anyway, if you can't reproduce it with this fix then it lgtm.

Reply:
Wait, this is not right i think, I did not test scaleup.  When scale up to multi-instances, it is different. The onAdd event could not contain the newly started gateway.